src/lib.rs

```diff
@@ -51,25 +51,39 @@ pub fn run(args: &clap::ArgMatches, current_dir: &PathBuf) -> Result<Vec<Warning
         };
 
         let mut lines = get_line_entries(&fe, strs);
+        let mut backup_file = None;
 
         // run fixers & write results to file
         let mut result = checks::run(&lines, &skip_checks);
         if is_fix && !result.is_empty() && fixes::run(&mut result, &mut lines, &skip_checks) > 0 {
             // create backup copy unless user specifies not to
             let should_backup = !args.is_present("no-backup");
             if should_backup {
-                let backup_file = fs_utils::backup_file(&fe)?;
-                println!("Original file was backed up to: {:?}\n", backup_file);
+                backup_file = Some(fs_utils::backup_file(&fe)?.into_os_string());
             }
 
             // write corrected file
             fs_utils::write_file(&fe.path, lines)?;
         }
 
-        warnings.extend(result);
+        let mode = if is_fix { Mode::Fix } else { Mode::Check };
+        outputs.push(Output::new(fe, backup_file, result, mode));
     }
 
-    Ok(warnings)
+    Ok(outputs)
+}
+
+/// Prints outputs with correct newlines.
+///
+/// Prints a newline after an output if it is not the last output or, for the
+/// last output, it has no warnings.
+pub fn print_outputs(outputs: Vec<Output>) {
```

Comment:
To tell the truth, I see that you should do some refactoring.
I sugger to create the `Output` struct, which will store all output and has this method.
In addition, I suggest separating the current `Output` structure into `Check` and `Fix`.
It will allow to get rid of `if/else` conditions in many files.
What do you think?

Reply:
This PR is already pretty big. I think we can do it later.